### DISCUSSION TITLE
what makes a library compatible with bb?

### DISCUSSION THREAD

**Discussion by irfn** [babashka/babashka#850] [Q&A]
If I find an incompatible library, what all can be done to make that library compatible.
Is there a checklist / test case of sorts which can give an idea about how much work is involved?
Is there any documentation on what all was done to make a particular library compatible.

**Comment by borkdude** [comment]
@irfn These are the main obstacles:

- Certain Java classes that are not available in bb (yet)
- Usage of `deftype` 
- Usage of `defrecord` with Java interfaces
- Usage of `reify` with more than one Java interface

**Reply by irfn** [reply]
Thanks @borkdude for your answer. 
Would you be able to tell which of these categories is `clojure.reflect` incompatibility?
I couldn't spot the usages that you mention so I am assuming its the Java classes that are missing. Is that correct?

**Reply by borkdude** [reply]
They key thing is probably here: https://github.com/clojure/clojure/blob/b0e5056217454073288e1643cd19e44999f081b8/src/clj/clojure/reflect.clj#L123

and then it loads https://github.com/clojure/clojure/blob/b0e5056217454073288e1643cd19e44999f081b8/src/clj/clojure/reflect/java.clj

which contains several classes that bb doesn't know about + at least one deftype.

**Reply by borkdude** [reply]
Btw, what is your use case for clojure.reflect in bb? The core function `bean` does work btw.

**Reply by irfn** [reply]
Thanks. I should have noticed that. 
I've been trying to use `bb` to drive the evaluations for org-mode babel. and it works very well.
Now I am trying to use more and more Jupyter style Clojure notebooks in Clojure and was trying to use `Incanter`.
And this is failing due to dependency on `clojure.reflect`
Ill stick with lightweight stuff on this mode of coding.

**Reply by borkdude** [reply]
@irfn Sometimes a few small tweaks go a long way. You can try forking that project and see how far you get.
